### DISCUSSION TITLE
Trouble with WaveShare USB to CAN-A

### DISCUSSION THREAD

**Discussion by ThanhMelb** [ai-republic/bms-to-inverter#64] [Q&A]
I'm new to Linux and I'm setting up Ubuntu with a USB-CAN-A interface. When I run the command ./canusb -t -d /dev/ttyUSB0 -s 500000 -t, my computer receives all the data from the CAN device, so the connection seems good. However, I’m unsure if I need to set up can0 for the BMS-to-inverter connection to work properly.

I'm using a Samsung BMS with the device set to /dev/ttyUSB0. I haven't encountered any errors, but the app isn't pulling data from the BMS.

Here’s the device I'm using: [Waveshare USB-CAN-A](https://www.waveshare.com/usb-can-a.htm).

Any help would be greatly appreciated!

**Comment by ai-republic** [comment]
Hi @ThanhMelb,
you do need a device mapping for the port to use. But your [Waveshare USB-CAN-A](https://www.waveshare.com/usb-can-a.htm) should create a USB device as shown on their Wiki.
![image](https://github.com/user-attachments/assets/ed7e6c90-6292-4e55-811d-f4f50d38d8bb)
Be sure to set the correct baud rate by calling something like:
```
sudo sudo stty -F /dev/ttyUSB0 500000
```
The baudrate should match your Samsung BMS CAN baudrate (default 500000).

**Reply by ThanhMelb** [reply]
Hi AI-Republic,

Thanks so much for your help! I've followed all the steps as recommended, but unfortunately, it's still not working. I must have done something wrong.

**Here’s my app config:**
bms.pollInterval=1
bms.1.type=SAMSUNG_CAN
bms.1.id=100
bms.1.portLocator=/dev/ttyUSB1
bms.1.baudRate=500000
bms.1.delayAfterNoBytes=200

**My USB is configured as USB1.**

root@raspberry:/opt/config# ls /dev/ttyU*
/dev/ttyUSB1

When I run root@raspberry:/opt/USB-CAN-A# ./canusb -t -d /dev/ttyUSB1 -s 500000 -t, I can see data coming from my BMS, so the cable seems to be connected properly:

>>> aa 55 12 03 01 00 00 00 00 00 00 00 00 00 01 00 00 00 00 17     '...............'
<<< aa c8 f0 05 01 00 45 0f 46 0f 45 0f 55
2200.444093 Frame ID: 05f0, Data: 0f 45 0f 46 0f 45 00 01
<<< aa c8 f1 05 01 00 46 0f 44 0f 43 0f 55
2200.445045 Frame ID: 05f1, Data: 0f 43 0f 44 0f 46 00 01

However, when I try running the app, it fails to open /dev/ttyUSB1:
2024-10-01 08:40:04.023 | INFO  | Thread-10  | toinverter.core.Port:111   | Opening /dev/ttyUSB1 ...
2024-10-01 08:40:04.024 | ERROR | Thread-10  | toinverter.core.Port:116   | Opening port /dev/ttyUSB1 FAILED!

Is there anything else I could try?

Again, thank you so much for your assistance!

**Comment by ai-republic** [comment]
@ThanhMelb could you attach you log file please, so I can view the whole initialization, etc.?

**Comment by ThanhMelb** [comment]
Hi ai-republic

I have no luck with linux, I did re-install the ubuntu as fresh install, now i can no-longer install the app, belo are error, 

nstalling in: /opt
Configuration in: /opt/config
Temp directory is: /opt/temp
Downloading application...done
Building application...
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] bms-to-inverter-parent                                             [pom]
[INFO] core-api-0.0.1-SNAPSHOT                                            [jar]
[INFO] protocol-can-0.0.1-SNAPSHOT                                        [jar]
[INFO] bms-byd-can-0.0.1-SNAPSHOT                                         [jar]
[INFO] inverter-deye-can-0.0.1-SNAPSHOT                                   [jar]
[INFO] webserver-0.0.1-SNAPSHOT                                           [jar]
[INFO] bms-to-inverter-main-0.0.1-SNAPSHOT                                [jar]
[INFO] 
[INFO] -------< com.ai-republic.bms-to-inverter:bms-to-inverter-parent >-------
[INFO] Building bms-to-inverter-parent 0.0.1-SNAPSHOT                     [1/7]
[INFO]   from pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ bms-to-inverter-parent ---
[INFO] 
[INFO] --------------< com.ai-republic.bms-to-inverter:core-api >--------------
[INFO] Building core-api-0.0.1-SNAPSHOT 0.0.1-SNAPSHOT                    [2/7]
[INFO]   from core-api/pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ core-api ---
[INFO] 
[INFO] --- resources:3.3.1:resources (default-resources) @ core-api ---
[INFO] Copying 1 resource from src/main/resources to target/classes
[INFO] 
[INFO] --- compiler:3.10.1:compile (default-compile) @ core-api ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 38 source files to /opt/temp/bms-to-inverter-main/core-api/target/classes
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for bms-to-inverter-parent 0.0.1-SNAPSHOT:
[INFO] 
[INFO] bms-to-inverter-parent ............................. SUCCESS [  0.205 s]
[INFO] core-api-0.0.1-SNAPSHOT ............................ FAILURE [  0.575 s]
[INFO] protocol-can-0.0.1-SNAPSHOT ........................ SKIPPED
[INFO] bms-byd-can-0.0.1-SNAPSHOT ......................... SKIPPED
[INFO] inverter-deye-can-0.0.1-SNAPSHOT ................... SKIPPED
[INFO] webserver-0.0.1-SNAPSHOT ........................... SKIPPED
[INFO] bms-to-inverter-main-0.0.1-SNAPSHOT ................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.910 s
[INFO] Finished at: 2024-10-04T20:27:34+10:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project core-api: Fatal error compiling: error: release version 17 not supported -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :core-api

Failed to build application!

Please help.
Thanks so much

**Comment by ai-republic** [comment]
Could you please try and download the latest _Configurator_ applications please?

**Reply by ThanhMelb** [reply]
Yeah, i have downloaded the latest one from github, i m going to re-build the box again.

Thanks so much

**Comment by ThanhMelb** [comment]
[BMS-to-Inverter.log](https://github.com/user-attachments/files/17318588/BMS-to-Inverter.log)

I successfully reinstalled Ubuntu, but I made a mistake with Java 21. After installing it, the configurator worked perfectly. I've attached the log file—could you please help me identify the problem with my setup? Thank you so much
ThanhMelb

**Reply by ai-republic** [reply]
Could you please deactivate the webserver in the configuration and run it again. The log get's flooded with the webserver stuff otherwise. (I really need to separate those log entries for the webserver 🤔)

**Comment by ThanhMelb** [comment]
Again, thanks for your help, i have attached the log and config file
[BMS-to-Inverter.log](https://github.com/user-attachments/files/17333467/BMS-to-Inverter.log)

###################################################################
###                  System specific settings                   ###
###################################################################

###################################################################
###                       BMS settings                          ###
###################################################################

####  Simple single port configuration ####
# bms.pollIntervall - is the interval to request BMS data (in seconds)
# bms.x.type - can be (DALY_CAN, DALY_RS485, JK_CAN, PYLON_CAN or SEPLOS_CAN 
# bms.x.portLocator - is the locator/device to use to communicate to the BMS, eg. can0, /dev/ttyUSB0, com3, etc.  
# bms.x.baudRate - is the locator/device baudrate to use to communicate to the BMS, eg. 9600, 500000, etc.  
# bms.x.delayAfterNoBytes - is the delay after receiving no data (in ms)
bms.pollInterval=1

bms.1.type=SAMSUNG_CAN
bms.1.id=1
bms.1.portLocator=/dev/ttyUSB0
bms.1.baudRate=500000
bms.1.delayAfterNoBytes=200

###################################################################
###                    Inverter settings                        ###
###################################################################
# The inverter type can be NONE, DEYE_CAN, GROWATT_CAN, SMA_SI_CAN, SOLARK_CAN
inverter.type=NONE
# The port name/device to use to communicate to the  inverter  
inverter.portLocator=can0
# The port baud rate to use to communicate to the  inverter  
inverter.baudRate=9600
# Interval to send data to the inverter (in seconds)
inverter.sendInterval=1

###################################################################
###                     Plugin settings                         ###
###################################################################
plugin.inverter.1.class=com.airepublic.bmstoinverter.core.plugin.inverter.PresetBatteryPackDataPlugin
plugin.inverter.1.property.1.name=SOC
plugin.inverter.1.property.1.value=100
plugin.inverter.1.property.1.description=The configured preset batterypack SOC (unit 0.1%)
plugin.inverter.1.property.2.name=SOH
plugin.inverter.1.property.2.value=990
plugin.inverter.1.property.2.description=The configured preset batterypack SOH (unit 0.1%)
plugin.inverter.1.property.3.name=Current
plugin.inverter.1.property.3.value=100
plugin.inverter.1.property.3.description=The configured preset batterypack current (unit 0.1A)
plugin.inverter.1.property.4.name=Voltage
plugin.inverter.1.property.4.value=48
plugin.inverter.1.property.4.description=The configured preset batterypack voltage (unit 0.1V)
plugin.inverter.1.property.5.name=Max. charge current
plugin.inverter.1.property.5.value=200
plugin.inverter.1.property.5.description=The configured preset batterypack maximum charge current (unit 0.1A)
plugin.inverter.1.property.6.name=Max. discharge current
plugin.inverter.1.property.6.value=200
plugin.inverter.1.property.6.description=The configured preset batterypack maximum discharge current (unit 0.1A)
plugin.inverter.1.property.7.name=Max. voltage limit
plugin.inverter.1.property.7.value=58
plugin.inverter.1.property.7.description=The configured preset batterypack maximum voltage limit (unit 0.1V)
plugin.inverter.1.property.8.name=Min. voltage lime
plugin.inverter.1.property.8.value=48
plugin.inverter.1.property.8.description=The configured preset batterypack minimum voltage limit (unit 0.1V)
plugin.inverter.1.property.9.name=Average Temperature
plugin.inverter.1.property.9.value=20
plugin.inverter.1.property.9.description=The configured preset batterypack average temperature (unit 0.1C)

plugin.inverter.2.class=com.airepublic.bmstoinverter.core.plugin.inverter.SimulatedBatteryPackPlugin
plugin.inverter.2.property.1.name=SOC
plugin.inverter.2.property.1.value=50
plugin.inverter.2.property.1.description=The configured preset batterypack SOC (unit 0.1%)
plugin.inverter.2.property.2.name=SOH
plugin.inverter.2.property.2.value=50
plugin.inverter.2.property.2.description=The configured preset batterypack SOH (unit 0.1%)
plugin.inverter.2.property.3.name=Current
plugin.inverter.2.property.3.value=10
plugin.inverter.2.property.3.description=The configured preset batterypack current (unit 0.1A)
plugin.inverter.2.property.4.name=Voltage
plugin.inverter.2.property.4.value=52
plugin.inverter.2.property.4.description=The configured preset batterypack voltage (unit 0.1V)
plugin.inverter.2.property.5.name=Max. charge current
plugin.inverter.2.property.5.value=100
plugin.inverter.2.property.5.description=The configured preset batterypack maximum charge current (unit 0.1A)
plugin.inverter.2.property.6.name=Max. discharge current
plugin.inverter.2.property.6.value=100
plugin.inverter.2.property.6.description=The configured preset batterypack maximum discharge current (unit 0.1A)
plugin.inverter.2.property.7.name=Max. voltage limit
plugin.inverter.2.property.7.value=58
plugin.inverter.2.property.7.description=The configured preset batterypack maximum voltage limit (unit 0.1V)
plugin.inverter.2.property.8.name=Min. voltage lime
plugin.inverter.2.property.8.value=44
plugin.inverter.2.property.8.description=The configured preset batterypack minimum voltage limit (unit 0.1V)
plugin.inverter.2.property.9.name=Average Temperature
plugin.inverter.2.property.9.value=25
plugin.inverter.2.property.9.description=The configured preset batterypack average temperature (unit 0.1C)

###################################################################
###                 Optional services settings                  ###
###################################################################

**Comment by ai-republic** [comment]
Ok, so in your configuration you do not need to activate any plugins. For now let's just see that you get data from your BMS.
So only have the _SAMSUNG_CAN_ BMS configured and nothing else. Please also use log level _debug_ for now just to see what data exactly is coming in. Thanks!

**Comment by ThanhMelb** [comment]
Hi,
I ran the app for an hour without encountering any errors or logs. I can see that the light on the USBCAN (ttyUSB0) is continuously flashing
![image](https://github.com/user-attachments/assets/f251b283-c54a-4658-81d3-e23167bff38d)
I also changed the baud rate to all different variable but no luck
Thanks

**Reply by ThanhMelb** [reply]
[ELPM482-00005 Manual.pdf](https://github.com/user-attachments/files/17359747/ELPM482-00005.Manual.pdf)
Attached is the documentation for the Samsung battery I have. I just want to confirm that this battery is supported by your software.
Thanks

**Comment by ai-republic** [comment]
Hi @ThanhMelb,
could you attach your config.properties, log4j2.xml and the log file _bmstoinverter.log_ please?
Yes, your Samsumg battery is supported as per feedback of other users (see [here](https://github.com/ai-republic/bms-to-inverter/issues/63)) 😄 
I suspect the problem is  some setup of your CAN adapter.

CC: @jezza61

**Reply by ThanhMelb** [reply]
[Config and Log Files.zip](https://github.com/user-attachments/files/17359863/Config.and.Log.Files.zip)
[BMS-to-Inverter.log](https://github.com/user-attachments/files/17359866/BMS-to-Inverter.log)Hi,

Thanks again for your help! I suspect there might be an issue with the type of USBCAN adapter I'm using or the OS version. I know you recommended the Raspberry CAN adapter, but do you have any suggestions for other adapters that work with your software?

Thanks so much!

**Comment by ai-republic** [comment]
This seems to be the problem:
![image](https://github.com/user-attachments/assets/0b2a82c9-6b18-4528-8d71-f7fbf8faf76a)
The CAN device cannot be found. I saw one user mapping the `/dev/ttyUSB0` to a network device somehow. Maybe that works.
Otherwise I would recommend the Waveshare CAN hat (either 2-CH CAN or RS485-CAN hat, depending on your inverter interface CAN/RS485)

Oh, and please do a _Clean install_ with the latest _Configurator_ 😉

**Reply by romanbetwo** [reply]
I have try mapping, still cant roll with waveshare CAN A USB stick, still getting the same error (and I bought 2)  so I do believe its because of that. going to try different OS .

**Comment by jezza61** [comment]
Hi,I have used canable usb adapter and also waveshare can-A and both have worked for me. Right now I am using a dual port waveshare can hat for raspberry. So far so good, now I am tuning the Inverter.That pesky rain does not help 😁
Best Regards
Jerry

**Reply by romanbetwo** [reply]
Hi Jerry, can you give me info about your setup? :D  Hw Sw?

**Comment by jezza61** [comment]
I used raspberry pi 4, with the latest 64 bit os.Then I installed Java and waveshare drivers.After that all was OK.Some waveshare drivers are meant to the older 32 bit os, so if you get stuck you can try that.I have tried raspberry 1b+  and 3 and these also worked. Best regardsJerrySent from my Galaxy
-------- Original message --------From: romanbetwo ***@***.***> Date: 15/10/24  7:01 am  (GMT+10:00) To: ai-republic/bms-to-inverter ***@***.***> Cc: jezza61 ***@***.***>, Mention ***@***.***> Subject: Re: [ai-republic/bms-to-inverter] Trouble with WaveShare USB to CAN-A (Discussion #64) 
Hi Jerry, can you give me info about your setup? :D  Hw Sw?

**Reply by romanbetwo** [reply]
can you be more specific with the drivers please?

**Comment by jezza61** [comment]
To be honest, it all depends on your skill.The easiest set-up would be with raspberry pi 4 and a waveshare can hat.You can pick the rs485/can version and that would cover you for all options.In my case I use raspberry 4, and a dual can hat.And with a very simple install it all works just fine.I have also tried pi 1+, but it was a bit slow.I used canable usb and a waveshare usb adapters, with canable being better. But I used these for testing only.So far my Samsung bms and deye inverter work just fine.However I have 3 different raspberry boxes connected.One runs solar assistant, and one node-red for my experimenting 
Best Regards 
Jerry

**Reply by romanbetwo** [reply]
Hi, yep understand. For me probably rs485+can hat would be a hit also but I want to make system versatile. Thats why I went with usb sticks. Currently I am trying to connect growatt sph 10000 which has Can bms (only) but I am stuck with /dev/ttyUSB port opening. You mention drivers so thats why I catch up :).

I will buy pi hat rs/can in this case and try to go with that.

Thanks

Roman

**Reply by ThanhMelb** [reply]
Thanks jezza61

I am going to buy 

https://www.waveshare.com/wiki/2-CH_CAN_HAT and 

https://raspberry.piaustralia.com.au/products/raspberry-pi-4?variant=16389894701105&currency=AUD&utm_medium=product_sync&utm_source=google&utm_content=sag_organic&utm_campaign=sag_organic&gad_source=1&gclid=Cj0KCQjwgrO4BhC2ARIsAKQ7zUmwmd6NVumrz9WRhEQpHX_sTOK9OtdKF9DVbKmJn5CSsmmZx9GblVwaAmEREALw_wcB

Do you think they will works?



Thanks so much

**Reply by ai-republic** [reply]
Yes, that's what I recommend if your BMS is CAN capable and also your inverter - you're best off with the 2CH-CAN hat which is also a lot more reliable than any USB adapter. 
If you need CAN on one side and RS485 on the other I would recommend the Waveshare RS485-CAN hat.

**Reply by romanbetwo** [reply]
![IMG_20241015_132635.jpg](https://github.com/user-attachments/assets/af98f1b1-1097-4ed0-97ee-323a58b32356)

So lets go :)

**Comment by jezza61** [comment]
That is what I use for the bms to inverter link.It works just fine.A bit of pain to load wave share drivers, as the instruction is for an older pi os, but it can be done.In my case I used the plugin to limit the top voltage for the battery.Right now I have the battery running from 10% to 85%.But I may possibly to adjust it later.I am running a separate raspberry 3 with rs485 usb adapter and solar assistant.Very nice interface and I can monitor from anywhere.I also have a raspberry 5 with node-red with a custom software in case I wanted to shut individual battery packs remotely, just on case.
Best regards
Jerry

**Comment by ai-republic** [comment]
Actually you do not need any of the Python demo applications. So you just need to follow how to install the hat and the Pi configuration from the Wiki. Check that the devices are shown as described in the Waveshare Wiki.
